# Machine Learning – Week 1

## Python Data Cleaning CLI Tool

### Introduction

This project is developed as part of the Machine Learning Week 1 curriculum. The main purpose of this project is to practice Python programming fundamentals, Pandas DataFrames, NumPy arrays, and basic data cleaning techniques.

The project focuses on creating a Python-based data cleaning tool that can load a local dataset, identify missing values, handle missing records, and save the cleaned dataset into a new file.

### Objective

The objective of this project is to:

- Practice Python data structures.
- Work with Pandas DataFrames.
- Perform basic numerical operations using NumPy.
- Identify and handle missing values in datasets.
- Save the cleaned dataset as a new CSV file.
- Understand the basic workflow of a data cleaning tool.

### Expected Outcome

At the end of this project, a cleaned dataset will be generated from the original dataset, and the data cleaning process will be documented.

## 2. Libraries

The following Python libraries are used in this project:

- **Pandas:** Used for loading, viewing, cleaning, and saving dataset files.
- **NumPy:** Used for numerical operations and array calculations.
- **Argparse:** Used to create the command-line interface (CLI) for the data cleaning tool.

These libraries provide the basic functionality required for data loading, data processing, and building the cleaning tool.

In [4]:
import pandas as pd
import numpy as np
import argparse

print("Pandas version:", pd.__version__)
print("NumPy version:", np.__version__)
print("Libraries loaded successfully!")

Pandas version: 2.2.3
NumPy version: 2.1.3
Libraries loaded successfully!


## 3. Python Data Structures

Python provides several built-in data structures for storing and organizing information.

The main data structures used in this project are:

- **List:** Stores multiple values in an ordered collection.
- **Tuple:** Stores multiple values in an ordered and immutable collection.
- **Dictionary:** Stores data in key-value pairs.
- **Set:** Stores unique values without duplicates.

These data structures are important for working with and manipulating data in Python.

In [5]:
# Python Data Structures

# 1. List
skills = ["Python", "Pandas", "NumPy"]
print("List:", skills)

# 2. Tuple
dataset_info = ("CSV", "Customer Data", 100)
print("Tuple:", dataset_info)

# 3. Dictionary
student = {
    "name": "Bhomika",
    "course": "Machine Learning",
    "week": 1
}
print("Dictionary:", student)

# 4. Set
numbers = {1, 2, 3, 3, 4}
print("Set:", numbers)

List: ['Python', 'Pandas', 'NumPy']
Tuple: ('CSV', 'Customer Data', 100)
Dictionary: {'name': 'Bhomika', 'course': 'Machine Learning', 'week': 1}
Set: {1, 2, 3, 4}


## 4. NumPy Arrays

NumPy is a Python library used for numerical operations and working with arrays.

In this project, NumPy is used to understand basic array operations such as creating arrays, calculating the mean, and performing mathematical operations.

NumPy arrays are useful when working with numerical data in machine learning and data analysis.

In [6]:
# NumPy Array Example

numbers = np.array([10, 20, 30, 40, 50])

print("NumPy Array:")
print(numbers)

print("\nMean:")
print(numbers.mean())

print("\nArray after multiplying by 2:")
print(numbers * 2)

print("\nMaximum value:")
print(numbers.max())

print("\nMinimum value:")
print(numbers.min())

NumPy Array:
[10 20 30 40 50]

Mean:
30.0

Array after multiplying by 2:
[ 20  40  60  80 100]

Maximum value:
50

Minimum value:
10


## 5. Dataset Loading

In this step, the dataset is loaded into a Pandas DataFrame.

Pandas provides the `read_csv()` function to read CSV files. After loading the dataset, we will inspect its rows, columns, data types, and missing values.

In [9]:
import pandas as pd

url = "https://raw.githubusercontent.com/rolandmueller/titanic/main/titanic3.csv"

df = pd.read_csv(url)

print("Dataset loaded successfully!")
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

df.head()

Dataset loaded successfully!
Rows: 1309
Columns: 14


,pclass,survived,name,sex,age,sibsp,parch,ticket,fare,cabin,embarked,boat,body,home.dest
0,1,1,"Allen, Miss. Elisabeth Walton",female,29.0000,0,0,24160,211.3375,B5,S,2,NaN,"St Louis, MO"
1,1,1,"Allison, Master. Hudson Trevor",male,0.9167,1,2,113781,151.5500,C22 C26,S,11,NaN,"Montreal, PQ / Chesterville, ON"
2,1,0,"Allison, Miss. Helen Loraine",female,2.0000,1,2,113781,151.5500,C22 C26,S,NaN,NaN,"Montreal, PQ / Chesterville, ON"
3,1,0,"Allison, Mr. Hudson Joshua Creighton",male,30.0000,1,2,113781,151.5500,C22 C26,S,NaN,135.0,"Montreal, PQ / Chesterville, ON"
4,1,0,"Allison, Mrs. Hudson J C (Bessie Waldo Daniels)",female,25.0000,1,2,113781,151.5500,C22 C26,S,NaN,NaN,"Montreal, PQ / Chesterville, ON"


## 6. Data Understanding

After loading the Titanic dataset, the next step is to understand its structure and contents.

In this step, we will examine the number of rows and columns, column names, data types, and missing values in the dataset.

This helps us understand the quality of the data before performing data cleaning.

In [10]:
print("Dataset Shape:")
print(df.shape)

print("\nColumn Names:")
print(df.columns.tolist())

print("\nData Types:")
print(df.dtypes)

Dataset Shape:
(1309, 14)

Column Names:
['pclass', 'survived', 'name', 'sex', 'age', 'sibsp', 'parch', 'ticket', 'fare', 'cabin', 'embarked', 'boat', 'body', 'home.dest']

Data Types:
pclass         int64
survived       int64
name          object
sex           object
age          float64
sibsp          int64
parch          int64
ticket        object
fare         float64
cabin         object
embarked      object
boat          object
body         float64
home.dest     object
dtype: object


In [11]:
print("Missing Values in Each Column:")
print(df.isnull().sum())

Missing Values in Each Column:
pclass          0
survived        0
name            0
sex             0
age           263
sibsp           0
parch           0
ticket          0
fare            1
cabin        1014
embarked        2
boat          823
body         1188
home.dest     564
dtype: int64


In [12]:
missing_percentage = (df.isnull().sum() / len(df)) * 100

print("Missing Value Percentage:")
print(missing_percentage)

Missing Value Percentage:
pclass        0.000000
survived      0.000000
name          0.000000
sex           0.000000
age          20.091673
sibsp         0.000000
parch         0.000000
ticket        0.000000
fare          0.076394
cabin        77.463713
embarked      0.152788
boat         62.872422
body         90.756303
home.dest    43.086325
dtype: float64


## 7. Save Clean Dataset

After cleaning the dataset, the cleaned data is saved as a new CSV file.

The original dataset is kept unchanged, while the cleaned dataset is exported separately for further use.

In [14]:
# Create a copy of the original dataset
clean_df = df.copy()

# Fill missing Age values with median
clean_df["age"] = clean_df["age"].fillna(clean_df["age"].median())

# Fill missing Fare values with median
clean_df["fare"] = clean_df["fare"].fillna(clean_df["fare"].median())

# Fill missing Embarked values with mode
clean_df["embarked"] = clean_df["embarked"].fillna(
    clean_df["embarked"].mode()[0]
)

print("Data cleaning completed!")

Data cleaning completed!


In [15]:
# Save cleaned dataset
output_file = "titanic_cleaned.csv"

clean_df.to_csv(output_file, index=False)

print("Cleaned dataset saved successfully!")
print("Output file:", output_file)

Cleaned dataset saved successfully!
Output file: titanic_cleaned.csv


In [16]:
print(clean_df.isnull().sum())

pclass          0
survived        0
name            0
sex             0
age             0
sibsp           0
parch           0
ticket          0
fare            0
cabin        1014
embarked        0
boat          823
body         1188
home.dest     564
dtype: int64


Step 8 — Save Clean Dataset

In [17]:
output_file = "titanic_cleaned.csv"

clean_df.to_csv(output_file, index=False)

print("Cleaned dataset saved successfully!")
print("File name:", output_file)

Cleaned dataset saved successfully!
File name: titanic_cleaned.csv


In [18]:
from google.colab import files

files.download("titanic_cleaned.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## 9. Python Data Cleaning CLI Tool

The data cleaning process is converted into a command-line Python script.

The CLI tool accepts an input CSV file and an output file path. It loads the input dataset, handles selected missing values, and saves the cleaned dataset.

In [21]:
%%writefile data_cleaner.py

import argparse
import pandas as pd


def clean_dataset(input_file, output_file):
    # Load dataset
    df = pd.read_csv(input_file)

    print("Dataset loaded successfully!")
    print("Original shape:", df.shape)

    # Create a copy
    clean_df = df.copy()

    # Handle missing numerical values
    if "age" in clean_df.columns:
        clean_df["age"] = clean_df["age"].fillna(clean_df["age"].median())

    if "fare" in clean_df.columns:
        clean_df["fare"] = clean_df["fare"].fillna(clean_df["fare"].median())

    # Handle missing categorical value
    if "embarked" in clean_df.columns:
        if not clean_df["embarked"].mode().empty:
            clean_df["embarked"] = clean_df["embarked"].fillna(
                clean_df["embarked"].mode()[0]
            )

    # Save cleaned dataset
    clean_df.to_csv(output_file, index=False)

    print("Cleaning completed successfully!")
    print("Cleaned shape:", clean_df.shape)
    print("Output file:", output_file)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Python Data Cleaning CLI Tool"
    )

    parser.add_argument(
        "--input",
        required=True,
        help="Path to input CSV file"
    )

    parser.add_argument(
        "--output",
        required=True,
        help="Path to output cleaned CSV file"
    )

    args = parser.parse_args()

    clean_dataset(args.input, args.output)

Writing data_cleaner.py


In [22]:
df.to_csv("titanic3.csv", index=False)

print("Original dataset saved as titanic3.csv")

Original dataset saved as titanic3.csv


In [23]:
!python data_cleaner.py --input titanic3.csv --output titanic_cli_cleaned.csv

Dataset loaded successfully!
Original shape: (1309, 14)
Cleaning completed successfully!
Cleaned shape: (1309, 14)
Output file: titanic_cli_cleaned.csv


In [24]:
import pandas as pd

cli_df = pd.read_csv("titanic_cli_cleaned.csv")

print(cli_df.shape)
print(cli_df.isnull().sum())

(1309, 14)
pclass          0
survived        0
name            0
sex             0
age             0
sibsp           0
parch           0
ticket          0
fare            0
cabin        1014
embarked        0
boat          823
body         1188
home.dest     564
dtype: int64


In [25]:
print("CLI Cleaning Result")
print("--------------------")
print("Rows:", cli_df.shape[0])
print("Columns:", cli_df.shape[1])

print("\nMissing Values:")
print(cli_df.isnull().sum())

CLI Cleaning Result
--------------------
Rows: 1309
Columns: 14

Missing Values:
pclass          0
survived        0
name            0
sex             0
age             0
sibsp           0
parch           0
ticket          0
fare            0
cabin        1014
embarked        0
boat          823
body         1188
home.dest     564
dtype: int64


## 10. CLI Cleaning Result

The Python CLI data cleaning tool successfully loaded the Titanic dataset and generated a cleaned CSV file.

The dataset contains 1309 rows and 14 columns. Missing values in the selected columns (`age`, `fare`, and `embarked`) were handled using median and mode imputation.

Columns with a very high percentage of missing values such as `cabin`, `boat`, `body`, and `home.dest` were not blindly imputed.

In [26]:
from google.colab import files

files.download("data_cleaner.py")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>